examples/ft-refund/solidity-contract/src/FtRefund.sol

```diff
@@ -0,0 +1,197 @@
+
+contract FtRefund is AccessControl {
+    using AuroraSdk for NEAR;
+    using AuroraSdk for PromiseCreateArgs;
+    using AuroraSdk for PromiseWithCallback;
+    using Codec for bytes;
+
+    bytes32 public constant CALLBACK_ROLE = keccak256("CALLBACK_ROLE");
+    bytes16 private constant _SYMBOLS = "0123456789abcdef";
+
+    IERC20 public wNEAR;
+    string public nearAccountId;
+    NEAR public near;
+
+    constructor(string memory _nearAccountId, IERC20 _wNEAR) {
+        nearAccountId = _nearAccountId;
+        near = AuroraSdk.initNear(_wNEAR);
+        wNEAR = _wNEAR;
+        _grantRole(
+            CALLBACK_ROLE,
+            AuroraSdk.nearRepresentitiveImplicitAddress(address(this))
+        );
+    }
+
+    // The purpose of this function is to get around a quirk of the XCC design.
+    // The `CALLBACK_ROLE` is logically the same as this contract because it is
+    // the implicit address of this contract's XCC account (and therefore all
+    // transactions from `CALLBACK_ROLE` actually originated from this contract).
+    // So this function is essentially just approving itself to spend its own tokens.
+    // The reason its needed is because the addresses are different on the level of
+    // the EVM even though they are logically the same entity as I have described above.
+    // To summarize, this function is needed because
+    // `nearRepresentitiveImplicitAddress(address(this)) != address(this)`, but we still
+    // want to spend wNEAR in Aurora callbacks.
+    function approveWNEAR() public {
+        uint256 amount = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+        PromiseCreateArgs memory approveCall = near.auroraCall(
+            address(this.wNEAR()),
+            abi.encodeWithSelector(
+                0x095ea7b3, // approve method selector
+                address(this),
+                amount
+            ),
+            0,
+            APPROVE_NEAR_GAS
+        );
+        approveCall.transact();
+    }
+
+    // Given an ERC-20 token that was bridged to Aurora from Near,
+    // this function will call the `ft_transfer_call` function on the
+    // Near-native NEP-141 token that ERC-20 was bridged from.
+    // This would be an important piece of an XCC-based integration into
+    // a Near app such as ref.finance (allowing Aurora users to use the
+    // Near app without creating a Near account themselves).
+    // For the purpose of this example, the call includes a message to
+    // force a refund back to the user because the point of this example
+    // is to show how to refund the Aurora user in the case there is a failure
+    // on the Near side.
+    // It is assumed that the sender has approved this contract to spend the
+    // sender's tokens.
+    function ftTransferCall(
+        IEvmErc20 token,
+        string memory tokenId,
+        uint128 amount
+    ) public {
+        token.transferFrom(msg.sender, address(this), amount);
+        token.withdrawToNear(
```

Comment:
Turns out what I thought was true about FIFO ordering was incorrect. The bridge did in fact encounter an error due to receipts being executed out of order. So the warning in the comment is accurate.

Fortunately the issue should come up rarely because there is still 1 block between when the `ft_transfer` receipt is created and when the `ft_transfer_call` receipt is. So the only way an error could happen is if there is congestion on the shard containing the NEP-141 token causing the `ft_transfer` receipt to be delayed by one or more blocks.

Reply:
This issue is addressed at the Engine level in https://github.com/aurora-is-near/aurora-engine/pull/868